### PR TITLE
compression: add brotli compressor and decompressor

### LINKED ISSUE #4429: Brotli Compression support in envoy
**Issue Template**

*Title*: *HTTP Compression Brotli support*

*Description*:
>Desired behavior 

- Envoy is capable for decompressing HTTP payload using Brotli compression/decompression algorithm

*Relevant Links*:
> https://github.com/google/brotli 
> https://quixdb.github.io/squash-benchmark/#results
> *https://github.com/google/brotli/tree/master/c*
@bmadhavan  @unicell @khgandhi @
/new-feature

### PR DESCRIPTION
**Pull request by rojkov**
Commit Message: compression: add brotli compressor and decompressor
Additional Description: Add new brotli compression extensions in addition to gzip.
Risk Level: Low, no existing functionality is touched
Testing: uni tests, manual tests with curl.
Docs Changes: updated docs for compression and decompression HTTP filters to refer the new available encoder/decoder.
Release Notes: updated current.rst
Fixes #4429

The PR adds a new dependency on https://github.com/google/brotli. Here's the current criteria answers:

| Criteria | Answer |
|---------|---------|
| Cloud Native Computing Foundation (CNCF) approved license | MIT |
| Dependencies must not substantially increase the binary size unless they are optional | brotli's binary size built with `-c opt` is 752K |
| No duplication of existing dependencies | no other dep provides Brotli |
| Hosted on a git repository and the archive fetch must directly reference this repository. | https://github.com/google/brotli |
| CVE history appears reasonable, no pathological CVE arcs | so far 4 CVEs related to brotli have been registered |
| Code review (ideally PRs) before merge | PRs are reviewed before merge |
| Security vulnerability process exists, with contact details and reporting/disclosure process | no policy exists, submitted https://github.com/google/brotli/issues/878 |
| > 1 contributor responsible for a non-trivial number of commits | 75 contributors |
| Tests run in CI | CI set up with AppVeyor and Github actions |
| High test coverage (also static/dynamic analysis, fuzzing) | Fuzzers are run in CI |
| Envoy can obtain advanced notification of vulnerabilities or of security releases | brotli is registered in CPE |
| Do other significant projects have shared fate by using this dependency? | Google Chrome is using the library |
| Releases (with release notes) | https://github.com/google/brotli/releases |
| Commits/releases in last 90 days | last commit 9 days ago |

### CODE REVIEW

File: source/extensions/compression/brotli/compressor/brotli_compressor_impl.h
Comment (jmarantz) [inline source/extensions/compression/brotli/compressor/brotli_compressor_impl.h:58]: `std::make_unique<uint8_t[]>(chunk_size)`

alternatively consider replacing this with MemBlockBuilder from source/common/common/mem_block_builder.h. I think it might be essentially the same functionality, though maybe not exactly drop-in.

**Review by jmarantz — COMMENTED** [review]
flushing comments

File: source/extensions/compression/brotli/compressor/brotli_compressor_impl.cc
Comment (rojkov) [reply source/extensions/compression/brotli/compressor/brotli_compressor_impl.cc]: Oops, that was a debug leftover copy-pasted in two places. Thank you!

File: source/extensions/compression/brotli/compressor/brotli_compressor_impl.cc
Comment (rojkov) [reply source/extensions/compression/brotli/compressor/brotli_compressor_impl.cc:45]: I stole this approach from the gzip encoder since it seems to consume a bit less memory. I guess gzip needs to be updated too.

File: source/extensions/compression/brotli/decompressor/brotli_decompressor_impl.h
Comment (rojkov) [reply source/extensions/compression/brotli/decompressor/brotli_decompressor_impl.h]: For now I've moved the struct to a common module. But probably it makes sense to turn it into a base class for the implementations in the similar fashion it's done for gzip. WDYT?

File: source/extensions/compression/brotli/compressor/brotli_compressor_impl.h
Comment (rojkov) [reply source/extensions/compression/brotli/compressor/brotli_compressor_impl.h]: Not sure MemBlockBuilder can be used here. We don't operate on this memory block. It's just passed to the Brotli lib.

Comment (jmarantz) [reply source/extensions/compression/brotli/compressor/brotli_compressor_impl.h]: Yeah you wouldn't use the append* functions but you can just pass it the block to the brotli lib using `span().data()`.

But I think I'll retract this suggestion because it looks like you need to pass pointers to the member variables to the Brotli library so we can't really abstract the buffer management, although inside the Brotli lib it's probably doing the same thing.

File: source/extensions/compression/brotli/compressor/brotli_compressor_impl.cc
Comment (jmarantz) [reply source/extensions/compression/brotli/compressor/brotli_compressor_impl.cc:45]: I wonder if it would be better to drain the buffer after every slice, so that the output buffer could potentially re-use the memory from the drained slices from the input buffer.

**Comment by junr03** [comment]
Thanks for the initial pass, will take a look today

**Review by junr03 — COMMENTED** [review]
initial pass. Super excited for this!

File: source/extensions/compression/brotli/compressor/brotli_compressor_impl.cc
Comment (rojkov) [reply source/extensions/compression/brotli/compressor/brotli_compressor_impl.cc:70]: For the compressor it must be an internal error or OOM. But in case of decompression it usually means the input data is garbage.

File: source/extensions/compression/brotli/decompressor/brotli_decompressor_impl.cc
Comment (rojkov) [reply source/extensions/compression/brotli/decompressor/brotli_decompressor_impl.cc:32]: The first loop runs until the input is not empty. But there might be situations when small input fully consumed by the decompressor unfolds into output not fitting the output chunk. Hence the second loop runs until there's no more output in the decompressor.

Would such a comment help?

File: source/extensions/compression/brotli/compressor/brotli_compressor_impl.cc
Comment (rojkov) [reply source/extensions/compression/brotli/compressor/brotli_compressor_impl.cc:45]: Thanks! Updated to drain the buffer as soon as possible.

File: api/envoy/extensions/compression/brotli/decompressor/v3/brotli.proto
Comment (rojkov) [reply api/envoy/extensions/compression/brotli/decompressor/v3/brotli.proto]: Agree, replaced with compressor/decompressor.

File: source/extensions/compression/brotli/compressor/brotli_compressor_impl.h
Comment (rojkov) [reply source/extensions/compression/brotli/compressor/brotli_compressor_impl.h:27]: Added a short comment about `default` being equal to `generic` currently in the brotli lib.

File: source/extensions/compression/brotli/compressor/brotli_compressor_impl.h
Comment (rojkov) [reply source/extensions/compression/brotli/compressor/brotli_compressor_impl.h]: Right. That's a leftover I missed once again from the times when there were two constructors. I dropped the comment. Thank you!

File: source/extensions/compression/brotli/compressor/brotli_compressor_impl.cc
Comment (rojkov) [reply source/extensions/compression/brotli/compressor/brotli_compressor_impl.cc]: I looked at Brotli's source code and it turned out `BrotliEncoderIsFinished()` internally just calls `BrotliEncoderHasMoreOutput()` and checks if the current operation is `FINISH`. So it's not really needed here and we should care only about the encoder's empty output.

File: source/extensions/compression/brotli/compressor/config.cc
Comment (rojkov) [reply source/extensions/compression/brotli/compressor/config.cc]: Added a clarification. But perhaps there is a better way to deal with zero values in omitted fields. I can only come up with making the field mandatory.

File: bazel/brotli.patch
Comment (htuch) [inline bazel/brotli.patch:10]: Can you please upstream this patch? We _heavily_ discourage custom patching during external dep fetch.

**Review by htuch — CHANGES_REQUESTED** [review]
/wait

File: bazel/brotli.patch
Comment (rojkov) [reply bazel/brotli.patch:10]: This patch has been already merged upstream: https://github.com/google/brotli/pull/839. Also https://github.com/google/brotli/pull/840. I put the references to the patch file as comments.

Comment (rojkov) [reply bazel/brotli.patch]: I've dropped the patch. Now the dep fetches from the main dev branch.

File: source/extensions/compression/brotli/compressor/brotli_compressor_impl.cc
Comment (rojkov) [reply source/extensions/compression/brotli/compressor/brotli_compressor_impl.cc]: Fixed it. Thank you!

**Review by jmarantz — APPROVED** [review]
@envoyproxy/senior-maintainers 

thank you, great work!

**Comment by rojkov** [comment]
@junr03 I think this is ready for another review round.

**Review by jmarantz — DISMISSED** [review]
Re-affirming my approval, and also wondering if we should follow up with a similar memory-trimming exercise in gzip.

**Comment by rojkov** [comment]
> also wondering if we should follow up with a similar memory-trimming exercise in gzip.

@jmarantz I'm working on that. Also I tried to avoid memcpy'ing in https://github.com/rojkov/envoy/commit/3035977d246301570a53c88b70a1806146abf2c7. But in my tests I haven't seen any noticeable improvement so far. I'm looking at nginx's code now.

File: source/extensions/compression/brotli/compressor/brotli_compressor_impl.cc
Comment (junr03) [reply source/extensions/compression/brotli/compressor/brotli_compressor_impl.cc:72]: Ah I see, so for the compressor you want to crash because there is nothing the user can do. But for the decompressor it could be user error (garbage input) so you want to alert via telemetry.

File: source/extensions/compression/brotli/decompressor/brotli_decompressor_impl.cc
Comment (junr03) [reply source/extensions/compression/brotli/decompressor/brotli_decompressor_impl.cc:32]: Yeah, inline comment would help. That way future readers can understand the interaction without needing to go to brotli source from the get go.

File: bazel/brotli.patch
Comment (junr03) [reply bazel/brotli.patch]: awesome

File: source/extensions/compression/brotli/compressor/config.cc
Comment (junr03) [reply source/extensions/compression/brotli/compressor/config.cc]: Ah, yeah you can use `google.protobuf.UInt32Value` so there is a difference between omitted and default.

File: source/extensions/compression/brotli/compressor/brotli_compressor_impl.cc
Comment (junr03) [reply source/extensions/compression/brotli/compressor/brotli_compressor_impl.cc]: inline comment here would also be good

**Review by junr03 — COMMENTED** [review]
Looking good! My only remaining comments are inline with threads we had before in the "unresolved comments".

File: source/extensions/compression/brotli/compressor/brotli_compressor_impl.cc
Comment (rojkov) [reply source/extensions/compression/brotli/compressor/brotli_compressor_impl.cc:77]: Yes, exactly. The gzip extensions has the same behavior.

File: source/extensions/compression/brotli/compressor/config.cc
Comment (rojkov) [reply source/extensions/compression/brotli/compressor/config.cc]: Oh, I didn't know that. Thanks! I dropped the decrement.

File: source/extensions/compression/brotli/compressor/brotli_compressor_impl.cc
Comment (rojkov) [reply source/extensions/compression/brotli/compressor/brotli_compressor_impl.cc]: done

File: source/extensions/compression/brotli/decompressor/brotli_decompressor_impl.cc
Comment (rojkov) [reply source/extensions/compression/brotli/decompressor/brotli_decompressor_impl.cc:32]: added a comment

**Review by junr03 — APPROVED** [review]
lgtm, maybe merge master for the asan failure?

**Comment by rojkov** [comment]
It still fails after merge... I fail to see how brotli can cause that except increased binary size of the tests.

**Comment by rojkov** [comment]
I can reproduce the failure locally. If I disable any other extension but brotli then the tests start to pass. Looks like the size of the test binary has reached some limit.

**Comment by jmarantz** [comment]
I haven't looked deeply, but is it possible to reorg the configs tests that are failing due to excessive size, so that we cover all the combinations across a few tests, but not with every extension loaded at the same time?

**Comment by rojkov** [comment]
Honestly I don't understand why the failing tests are linked with all the extensions in first place. It wasn't always this way: the dependency was added in https://github.com/envoyproxy/envoy/pull/2906/files#diff-ad37147ad0ed3649f7179f88851c7df0

@mattklein123 do you remember why it's needed?

After I had removed it the tests started to pass.

**Comment by rojkov** [comment]
Sorry, disregard my previous comment. It's only `//test/config_test:deprecated_configs_test` that passes without the dep. But `//test/config_test/example_configs_test` does need it.

**Comment by rojkov** [comment]
PTAL.
Finally all tests pass again!

File: test/config_test/BUILD
Comment (rojkov) [inline test/config_test/BUILD:53]: @htuch May I ask you what would be a better solution here? The problem is that the tests `//test/config_test/...` fail when the brotli (de)compressor is added to `envoy_all_extensions()` and ASan instrumentalization is added to the test binaries (presumably due to excessive size of the code segments). I kinda solved the problem by linking `//test/config_test:example_configs_test` with those extensions only that are used in the example configs. And `//test/config_test:deprecated_configs_test` doesn't seem to depend on `envoy_all_extensions()` at all.

But I suspect that the tool `//test/tools/config_load_check:config_load_check_tool` is useless now. I tried to link the tool with `envoy_all_extensions()`, but stumbled upon visibility constraints:
```
ERROR: /source/test/tools/config_load_check/BUILD:14:22: in cc_library rule //test/tools/config_load_check:config_load_check_lib: target '//source/extensions/internal_redirect/allow_listed_routes:config' is not visible from target '//test/tools/config_load_check:config_load_check_lib'. Check the visibility declaration of the former target if you think the dependency is legitimate
```

**Comment by rojkov** [comment]
@htuch PTAL

**Comment by htuch** [comment]
@rojkov I'm going to have to defer API review to another @envoyproxy/api-shepherds on this one as I'm going to be OOO for a bit.

**Comment by rojkov** [comment]
@lizan Perhaps you can check the API changes?

File: api/envoy/extensions/compression/brotli/compressor/v3/brotli.proto
Comment (htuch) [inline api/envoy/extensions/compression/brotli/compressor/v3/brotli.proto:24]: Where do these come from? How will users decide which to use?

Comment (htuch) [reply api/envoy/extensions/compression/brotli/compressor/v3/brotli.proto:24]: I see, you comment below on this.

**Review by htuch — COMMENTED** [review]
/lgtm api

**Comment by rojkov** [comment]
@moderation could you please lgtm deps?

**Comment by moderation** [comment]
@rojkov there are two pending feedback comments on deps. One is formatting and one is a correct UTC date

**Comment by rojkov** [comment]
@moderation Thanks! I've corrected the date, but what was the comment about formatting? Can't find it.

**Review by moderation — CHANGES_REQUESTED** [review]

File: bazel/repository_locations.bzl
Comment (moderation) [reply bazel/repository_locations.bzl:379]: You should be able to see it now - https://github.com/envoyproxy/envoy/pull/12998#discussion_r520121517

Comment (rojkov) [reply bazel/repository_locations.bzl]: Thank you! Updated now.

**Comment by rojkov** [comment]
@junr03 @jmarantz could you please give it another look. The API and deps approvals are in place now.

**Review by junr03 — APPROVED** [review]
re-affirming my earlier approval now that everything is green!

**Review by jmarantz — APPROVED** [review]
reaffirming approval

File: api/envoy/extensions/compression/brotli/compressor/v3/brotli.proto
Comment (rojkov) [reply api/envoy/extensions/compression/brotli/compressor/v3/brotli.proto:36]: I agree. Saner defaults wouldn't harm definitely. What would be the best choices for your use cases?

Comment (SaveTheRbtz) [reply api/envoy/extensions/compression/brotli/compressor/v3/brotli.proto:36]: Looking at cloudflare they use `WSIZE`=18 (262128) and compression level somewhere between 3 and 4 (based on `brotlidump.py`).  This seems relatively sane for dynamic content.  You may consider cracking down on WSIZE even more to bring it closer to gzip.

**Comment by junr03** [comment]
@envoyproxy/senior-maintainers can we get one of you to do a last pass here?

File: source/extensions/extensions_build_config.bzl
Comment (lizan) [inline source/extensions/extensions_build_config.bzl:27]: nit: indent

File: api/envoy/extensions/compression/brotli/compressor/v3/brotli.proto
Comment (rojkov) [reply api/envoy/extensions/compression/brotli/compressor/v3/brotli.proto:36]: Thanks! Updated the defaults.

File: source/extensions/extensions_build_config.bzl
Comment (rojkov) [reply source/extensions/extensions_build_config.bzl]: Fixed

**Comment by rojkov** [comment]
@envoyproxy/api-shepherds @envoyproxy/dependency-shepherds could you give your approval once again?

**Comment by moderation** [comment]
/lgtm deps

Latest commit is fcda9db7fd554ffb19c2410b9ada57cdabd19de5 with `release_date` of `2020-10-08` if you wanted to bump to the latest. We can bump in a followup or wait until next release

**Review by htuch — COMMENTED** [review]
/lgtm api

**Comment by junr03** [comment]
@lizan do you mind doing a final pass?

**Comment by rojkov** [comment]
Merged the latest master to resolve a merge conflict in `current.rst`.

**Comment by rojkov** [comment]
@lizan could you take a look?

**Review by htuch — COMMENTED** [review]
/lgtm deps

Thanks so much for the detailed scorecard!

**Review by lizan — APPROVED** [review]

**Comment by moderation** [comment]
I'm getting build errors as there isn't a `category` specified in https://github.com/envoyproxy/envoy/blob/main/source/extensions/compression/brotli/compressor/BUILD#L24-L35

```
ERROR: Traceback (most recent call last):
        File "~/Library/envoyproxy/envoy/source/extensions/compression/brotli/compressor/BUILD", line 24, column 19, in <toplevel>
                envoy_cc_extension(
        File "~/Library/envoyproxy/envoy/bazel/envoy_library.bzl", line 124, column 13, in envoy_cc_extension
                fail("Category not set for %s" % name)
Error in fail: Category not set for config
ERROR: Traceback (most recent call last):
        File "~/Library/envoyproxy/envoy/source/extensions/compression/brotli/decompressor/BUILD", line 26, column 19, in <toplevel>
                envoy_cc_extension(
        File "~/Library/envoyproxy/envoy/bazel/envoy_library.bzl", line 124, column 13, in envoy_cc_extension
                fail("Category not set for %s" % name)
Error in fail: Category not set for config
ERROR: ~/Library/envoyproxy/envoy/source/exe/BUILD:28:17: no such target '//source/extensions/compression/brotli/compressor:config_envoy_extension': target 'config_envoy_extension' not declared in package 'source/extensions/compression/brotli/compressor' defined by ~/Library/envoyproxy/envoy/source/extensions/compression/brotli/compressor/BUILD and referenced by '//source/exe:envoy_common_lib'
ERROR: ~/Library/envoyproxy/envoy/source/exe/BUILD:28:17: no such target '//source/extensions/compression/brotli/decompressor:config_envoy_extension': target 'config_envoy_extension' not declared in package 'source/extensions/compression/brotli/decompressor' defined by ~/Library/envoyproxy/envoy/source/extensions/compression/brotli/decompressor/BUILD and referenced by '//source/exe:envoy_common_lib'
ERROR: Analysis of target '//source/exe:envoy-static.stripped' failed; build aborted: Analysis failed
INFO: Elapsed time: 0.842s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded, 0 targets configured)
```

**Comment by moderation** [comment]
```diff
diff --git a/source/extensions/compression/brotli/compressor/BUILD b/source/extensions/compression/brotli/compressor/BUILD
index efd1739e0..93f13d8fe 100644
--- a/source/extensions/compression/brotli/compressor/BUILD
+++ b/source/extensions/compression/brotli/compressor/BUILD
@@ -25,6 +25,7 @@ envoy_cc_extension(
     name = "config",
     srcs = ["config.cc"],
     hdrs = ["config.h"],
+    category = "envoy.compression.decompressor",
     security_posture = "robust_to_untrusted_downstream",
     deps = [
         ":compressor_lib",
diff --git a/source/extensions/compression/brotli/decompressor/BUILD b/source/extensions/compression/brotli/decompressor/BUILD
index 3cc016e2f..3667300a8 100644
--- a/source/extensions/compression/brotli/decompressor/BUILD
+++ b/source/extensions/compression/brotli/decompressor/BUILD
@@ -27,6 +27,7 @@ envoy_cc_extension(
     name = "config",
     srcs = ["config.cc"],
     hdrs = ["config.h"],
+    category = "envoy.compression.decompressor",
     security_posture = "robust_to_untrusted_downstream",
     deps = [
         ":decompressor_lib",
```

**Pull request merged**
